File: src/com/cattailsw/nanidroid/shiori/SatoriPosixShiori.java
package com.cattailsw.nanidroid.shiori;

import java.nio.charset.Charset;

public class SatoriPosixShiori extends JNIShiori {
    static {
	System.loadLibrary("satoriya");
    }

    private String path = null;

    public SatoriPosixShiori(String p) {
	path = p;

	load(p);
    }

    public String request(String req) {
	String tmp_r = req.substring(0, req.length() - 2) + "Charset: Shift_JIS\r\n\r\n";
	byte[] result;
	try {
	    result = requestFromJNI2( tmp_r.getBytes(Charset.forName("Shift_JIS")) );
	}
	catch(Exception e) {
	    e.printStackTrace();
	    return NanidroidShiori.RES_NO_CONTENT;
	}
	
	return modResponseWithCharSet(result);// super.request( tmp_r );
    }

    public void unloadShiori(){
	unload();
    }

    public native void load(String path);
    public native void unload();
    public native byte[] requestFromJNI2(byte[] req);
}

File: src/com/cattailsw/nanidroid/shiori/SatoriPosixShiori.kt
package com.cattailsw.nanidroid.shiori

import java.nio.charset.Charset

class SatoriPosixShiori(private val path: String) : JNIShiori() {
    init {
        System.loadLibrary("satoriya")
        load(path)
    }

    override fun request(request: String): String {
        val rewritten = request.substring(0, request.length - 2) + "Charset: Shift_JIS\r\n\r\n"
        return try {
            modResponseWithCharSet(requestFromJNI2(rewritten.toByteArray(Charset.forName("Shift_JIS"))))
        } catch (error: Exception) {
            error.printStackTrace()
            NanidroidShiori.RES_NO_CONTENT
        }
    }

    override fun unloadShiori() { unload() }

    external fun load(path: String)
    external fun unload()
    external fun requestFromJNI2(req: ByteArray): ByteArray
}
